File: arrays/kadane_and_like/max_sum_rect_no_larger_than_k.cpp
// The naive solution is brute-force, which is O((mn)^2). In order to be more efficient,
// I tried something similar to Kadane's algorithm. The only difference is that here we have upper bound restriction K.
// Here's the easily understanding video link for the problem "find the max sum rectangle in 2D array": 
// Maximum Sum Rectangular Submatrix in Matrix dynamic programming/2D kadane (Trust me, it's really easy and straightforward).

// Once you are clear how to solve the above problem, the next step is to find the max sum no more than K in an array. This can be done within O(nlogn), and you can refer to this article: max subarray sum no more than k.
// For the solution below, I assume that the number of rows is larger than the number of columns. Thus in general time complexity is O[min(m,n)^2 * max(m,n) * log(max(m,n))], space O(max(m, n)).

int maxSumSubmatrix(vector<vector<int>>& matrix, int k) {
    if (matrix.empty()) return 0;
    int row = matrix.size(), col = matrix[0].size(), res = INT_MIN;
    for (int l = 0; l < col; ++l) {
        vector<int> sums(row, 0);
        for (int r = l; r < col; ++r) {
            for (int i = 0; i < row; ++i) {
                sums[i] += matrix[i][r];
            }
            
            // Find the max subarray no more than K 
            set<int> accuSet;
            accuSet.insert(0);
            int curSum = 0, curMax = INT_MIN;
            for (int sum : sums) {
                curSum += sum;
                set<int>::iterator it = accuSet.lower_bound(curSum - k);
                if (it != accuSet.end()) curMax = std::max(curMax, curSum - *it);
                accuSet.insert(curSum);
            }
            res = std::max(res, curMax);
        }
    }
    return res;
}

File: arrays/kadane_and_like/max_sum_rect_no_larger_than_k.py
# import bisect
# import sys
# class Solution(object):
#     def maxSumSubmatrix(self, matrix, k):
#         """
#         :type matrix: List[List[int]]
#         :type k: int
#         :rtype: int
#         """
#         if len(matrix) == 0:
#             return 0
#         m, n, res = len(matrix), len(matrix[0]), float("-inf")
#         for left in range(n):
#             sums = [0] * m

#             for right in range(left, n):
#                 for i in range(m):
#                     sums[i] += matrix[i][right]
#                 treeSet = [0]
#                 curSum = 0

#                 for sum in sums:
#                     curSum += sum
#                     i = bisect.bisect_left(treeSet, curSum - k)
#                     if i != len(treeSet):
#                         res = max(res, curSum - treeSet[i])
#                     bisect.insort(treeSet, curSum)
                    
#         return res

import bisect
import sys

class Solution(object):
    def max_sum_no_more_than_k(self, a, k):
        cum_list = [0]
        cum = 0
        max_so_far = -sys.maxint

        for i in range(len(a)):
            cum += a[i]
            cum_j_index = bisect.bisect_left(cum_list, cum - k)  # if bisect_left, no larger than k
            if cum_j_index < len(cum_list):
                max_so_far = max(max_so_far, cum - cum_list[cum_j_index])
            bisect.insort(cum_list, cum)

        return max_so_far

    def maxSumSubmatrix(self, matrix, k):
        if len(matrix)==0:
            return 0

        row_num = len(matrix)
        max_so_far = -sys.maxint
        columns = zip(*matrix)

        for l in range(len(columns)):
            current = [0]*row_num
            for colunm in columns[l:]:
                current = map(int.__add__, current, colunm)
                ans = self.max_sum_no_more_than_k(current, k)
                max_so_far = max(max_so_far, ans)

        return max_so_far
